WordPress/gutenberg
# Create withAPIData higher-order component for managing API data

Related: #902

This pull request seeks to introduce a new `withAPIData` higher-order component to provide a simple interface for fetching, creating, updating, or deleting data from the REST API.

Out of the box, it includes:

- Auto-fetching when your component mounts
- Reusing cached data if request has already been made
- Provides status updates so you can render accordingly
- Trigger creation, updates, or deletes on data

Example:

```jsx
function MyPost( { post } ) {
	if ( post.isLoading ) {
		return <div>Loading...</div>;
	}

	return <div>{ post.data.title.rendered }</div>;
}

export default withAPIData( ( props, { type } ) => ( {
	post: `/wp/v2/${ type( 'post' ) }/${ props.postId }`
} ) )( MyPost );
```

__Implementation notes:__

`withApiData` behaves similar to, and in some cases complements, React-Redux's `connect` function. It accepts props which can either be passed from the parent component, or composed from another higher-order component (e.g. `connect`), and returns an object of propName -> endpoint result.

Data-bound props take the shape of an object with a number of properties, depending on the methods supported for the particular endpoint:

- `GET`
   - `isLoading`: Whether the resource is currently being fetched
   - `data`: The resource, available once fetch succeeds
   - `get`: Function to invoke a new fetch request
   - `error`: The error object, if the fetch failed
- `POST`
   - `isCreating`: Whether the resource is currently being created
   - `createdData`: The created resource, available once create succeeds
   - `create`: Function to invoke a new create request
   - `createError`: The error object, if the create failed
- `PUT`
   - `isSaving`: Whether the resource is currently being saved
   - `savedData`: The saved resource, available once save succeeds
   - `save`: Function to invoke a new save request
   - `saveError`: The error object, if the save failed
- `PATCH`
   - `isPatching`: Whether the resource is currently being patched
   - `patchedData`: The patched resource, available once patch succeeds
   - `patch`: Function to invoke a new patch request
   - `patchError`: The error object, if the patch failed
- `DELETE`
   - `isDeleting`: Whether the resource is currently being deleted
   - `deletedData`: The deleted resource, available once delete succeeds
   - `delete`: Function to invoke a new delete request
   - `deleteError`: The error object, if the delete failed

There is some inspiration here from [Heroku's `react-refetch` project](https://github.com/heroku/react-refetch).

__Testing instructions:__

Verify that there are no regressions in the LastRevision component (requires post with at least one revision to appear).

---

_See original pull request comment: https://gist.github.com/aduth/e5dd6327963475e9f580b8ece523ffa0_
